compiler/rustc_typeck/src/collect.rs

```diff
@@ -857,7 +857,21 @@ fn convert_trait_item(tcx: TyCtxt<'_>, trait_item_id: hir::TraitItemId) {
             // Account for `const C: _;`.
             let mut visitor = HirPlaceholderCollector::default();
             visitor.visit_trait_item(trait_item);
-            placeholder_type_error(tcx, None, visitor.0, false, None, "constant");
+            placeholder_type_error(
+                tcx,
+                None,
+                visitor
+                    .0
+                    .into_iter()
+                    // When the const does not have a type, this error should not be emitted.
+                    .filter(|span| {
+                        !tcx.sess.diagnostic().has_stashed_diagnostic(*span, StashKey::ItemNoType)
+                    })
+                    .collect::<Vec<_>>(),
+                false,
+                None,
+                "constant",
+            );
         }
```

Comment:
This is a bit more generalized than necessary -- `ItemNoType` will only be stashed if we have a `_` infer type as the type of the const -- it will never be nested. Try something like:

```rust
hir:TraitItemKind(hir_ty, ..) => {
  tcx.ensure().type_of(trait_item_id.def_id);
  // Account for `const C: _;`.
  let mut visitor = HirPlaceholderCollector::default();
  visitor.visit_trait_item(trait_item);
  if !tcx.sess.diagnostic().has_stashed_diagnostic(hir_ty.span, StashKey::ItemNoType) {
    placeholder_type_error(tcx, None, visitor.0, false, None, "constant");
  }
}
```